### PR TITLE
Homework5

### PR DESCRIPTION
**Pull request by AntonMurashov**

### CODE REVIEW

**Comment by AinuraAliieva** [comment]
@AntonMurashov  Please merge master into the current branch

**Comment by AntonMurashov** [comment]
Done!

File: src/app/services/course.service.ts [hunk 1]
```diff
@@ -51,9 +51,27 @@ export class CourseService {
...
+  
 
   public findAll(): ICourse[] {
-    return this.getMockCourses();
+    return CourseService.mockCourses;
+  }
+
+  public createCourse(): ICourse[] {
+    console.log('AddCourse clicked');
+    return CourseService.mockCourses;
```
[the remark this answers is not in the document]
Comment (AntonMurashov) [reply src/app/services/course.service.ts]: Fixed

File: src/app/services/authorization.service.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  public Logout() {
+    this.userName = '';
+    this.token = '';
+    this.checkAuth.next({
+      isAuthentificated: this.IsAuthentificated(),
+      userName: this.GetUserInfo()
+    });
+  }
+}
```
[the remark this answers is not in the document]
Comment (AntonMurashov) [reply src/app/services/authorization.service.ts]: No benefits... fixed

File: src/app/services/course.service.ts [hunk 3]
```diff
@@ -51,9 +51,27 @@ export class CourseService {
...
+    CourseService.mockCourses = CourseService.mockCourses.filter((course: ICourse) => course.id !== id);
+    return CourseService.mockCourses;
+  }
+
+  public getCourse(id: number): ICourse {
+    return CourseService.mockCourses.find((course: ICourse) => course.id == id);
+  }
+
+  public updateCourse(id: number) {
+    console.log('Updating course ' + id);
```
[the remark this answers is not in the document]
Comment (AntonMurashov) [reply src/app/services/course.service.ts]: Fixed, but there is no logic for updating in courses component by now

File: src/app/services/authorization.service.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    return this.token != '';
+  }
+
+  public GetUserInfo(): string {
+    return this.userName;
+  }
+
+  public Login(email: string, password: string) {
+    this.userName = 'Test User' + ' (' + email + ')';
+    this.token = 'testtoken';
```
[the remark this answers is not in the document]
Comment (AntonMurashov) [reply src/app/services/authorization.service.ts]: Didn't notice that requirement :(  Fixed

File: src/app/core/header/header.component.html [hunk 5]
```diff
@@ -2,11 +2,11 @@
...
     </div>
-    <div class="profile user">
-        <img class="user" src="assets/img/ic-user.svg">
-        User login
+    <div class="profile user" *ngIf="isAuth">
+        <img src="assets/img/ic-user.svg">
+        {{userName}}
     </div>
-    <div class="profile" (click)="exit()">
+    <div class="profile" (click)="exit()" *ngIf="isAuth">
```
Comment (AinuraAliieva) [inline src/app/core/header/header.component.html]: I suggest optimizing this part by using ng-container with one subscription.
In that case, we reduce subscriptions from two to one :)

something like that
```
<ng-container *ngIf="isAuth">
  <div class="profile user"> ...
  <div class="profile" (click)="exit()"> ...
```

File: src/app/app.component.ts [hunk 6]
```diff
@@ -1,9 +1,23 @@
...
+  providers: [AuthorizationService]
 })
 export class AppComponent {
+
+  private isAuth = false;
+  _subscription: Subscription;
+
+  constructor(private authService: AuthorizationService) {
+    this.isAuth = this.authService.isAuthentificated();
+    this._subscription = authService.checkAuth.subscribe((value) => {
```
Comment (AinuraAliieva) [inline src/app/app.component.ts]: To increase performance I suggest unsubscribing on destroy phase

File: src/app/core/header/header.component.ts [hunk 7]
```diff
@@ -7,12 +9,22 @@ import { Component, OnInit } from '@angular/core';
...
+  private isAuth = false;
+  private userName = '';
+
+  constructor(private authService: AuthorizationService) {
+    this.userName = this.authService.getUserInfo();
+    this.isAuth = this.authService.isAuthentificated();
+    authService.checkAuth.subscribe(value => { 
+      this.isAuth = value.isAuthentificated; 
+      this.userName = value.userName;
+    });
```
Comment (AinuraAliieva) [inline src/app/core/header/header.component.ts]: Please save this subscription in some property and then unsubscribe on destroy phase to increase performance

**Comment by AntonMurashov** [comment]
Good points - all fixed!

**Pull request merged**